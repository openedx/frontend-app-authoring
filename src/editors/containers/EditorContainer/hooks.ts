import { useState } from 'react';
import { useSelector } from 'react-redux';

import analyticsEvt from '../../data/constants/analyticsEvt';
import { RequestKeys } from '../../data/constants/requests';
import { selectors } from '../../data/redux';
import { StrictDict } from '../../utils';
import * as appHooks from '../../hooks';

export const {
  clearSaveError,
  navigateCallback,
  nullMethod,
  saveBlock,
} = appHooks;

export const state = StrictDict({
  // eslint-disable-next-line react-hooks/rules-of-hooks
  isCancelConfirmModalOpen: (val) => useState(val),
});

export const handleSaveClicked = ({
  dispatch,
  getContent,
  validateEntry,
  returnFunction,
}) => {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  const returnUrl = useSelector(selectors.app.returnUrl);
  const destination = returnFunction ? '' : returnUrl;
  // eslint-disable-next-line react-hooks/rules-of-hooks
  const analytics = useSelector(selectors.app.analytics);

  return () => saveBlock({
    analytics,
    content: getContent({ dispatch }),
    destination,
    dispatch,
    returnFunction,
    validateEntry,
  });
};

export const cancelConfirmModalToggle = () => {
  const [isCancelConfirmOpen, setIsOpen] = state.isCancelConfirmModalOpen(false);
  return {
    isCancelConfirmOpen,
    openCancelConfirmModal: () => setIsOpen(true),
    closeCancelConfirmModal: () => setIsOpen(false),
  };
};

export const handleCancel = ({
  onClose = null,
  returnFunction = null,
}: {
  onClose?: (() => void) | null;
  returnFunction?: (() => (result: any) => void) | null;
}): ((result?: any) => void) => {
  if (onClose) {
    return onClose;
  }
  // eslint-disable-next-line react-hooks/rules-of-hooks
  const returnUrl = useSelector(selectors.app.returnUrl);
  return navigateCallback({
    returnFunction,
    // eslint-disable-next-line react-hooks/rules-of-hooks
    destination: returnFunction ? '' : returnUrl,
    analyticsEvent: analyticsEvt.editorCancelClick,
    // eslint-disable-next-line react-hooks/rules-of-hooks
    analytics: useSelector(selectors.app.analytics),
  });
};

// eslint-disable-next-line react-hooks/rules-of-hooks
export const isInitialized = () => useSelector(selectors.app.isInitialized);

// eslint-disable-next-line react-hooks/rules-of-hooks
export const saveFailed = () => useSelector((rootState) => (
  selectors.requests.isFailed(rootState, { requestKey: RequestKeys.saveBlock })
));
