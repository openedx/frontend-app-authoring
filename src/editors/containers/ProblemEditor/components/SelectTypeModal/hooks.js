import { useEffect } from 'react';
import {
  AdvanceProblemKeys, AdvanceProblems, ProblemTypeKeys, ProblemTypes,
} from '../../../../data/constants/problem';
import { snakeCaseKeys } from '../../../../utils';
import { getDataFromOlx } from '../../../../data/redux/thunkActions/problem';

export const onSelect = ({
  selected,
  updateField,
  setBlockTitle,
  defaultSettings,
}) => () => {
  if (Object.values(AdvanceProblemKeys).includes(selected)) {
    updateField({ problemType: ProblemTypeKeys.ADVANCED, rawOLX: AdvanceProblems[selected].template });
    setBlockTitle(AdvanceProblems[selected].title);
  } else {
    const newOLX = ProblemTypes[selected].template;
    const newState = getDataFromOlx({
      rawOLX: newOLX,
      rawSettings: {
        weight: 1,
        attempts_before_showanswer_button: 0,
        show_reset_button: null,
        showanswer: null,
        defaultToAdvanced: false,
      },
      defaultSettings: snakeCaseKeys(defaultSettings),
    });
    updateField(newState);
    setBlockTitle(ProblemTypes[selected].title);
  }
};

export const useArrowNav = (selected, setSelected) => {
  const detectKeyDown = (e) => {
    const problemTypeValues = Object.values(ProblemTypeKeys);
    switch (e.key) {
      case 'ArrowUp':
        if (problemTypeValues.includes(selected) && ProblemTypes[selected].prev) {
          setSelected(ProblemTypes[selected].prev);
          document.getElementById(ProblemTypes[selected].prev).focus();
        }
        break;
      case 'ArrowDown':
        if (problemTypeValues.includes(selected) && ProblemTypes[selected].next) {
          setSelected(ProblemTypes[selected].next);
          document.getElementById(ProblemTypes[selected].next).focus();
        }
        break;
      default:
    }
  };
  useEffect(() => {
    document.addEventListener('keydown', detectKeyDown, true);
    return () => {
      document.removeEventListener('keydown', detectKeyDown, true);
    };
  }, [selected, setSelected]);
};
