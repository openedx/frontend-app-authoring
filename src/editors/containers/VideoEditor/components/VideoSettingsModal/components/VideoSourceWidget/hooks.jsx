import React from 'react';
import { actions } from '../../../../../../data/redux';
import { parseYoutubeId } from '../../../../../../data/services/cms/api';
import * as requests from '../../../../../../data/redux/thunkActions/requests';

export const state = {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  showVideoIdChangeAlert: (args) => React.useState(args),
};

export const sourceHooks = ({ dispatch, previousVideoId, setAlert }) => ({
  updateVideoURL: (e, videoId) => {
    const videoUrl = e.target.value;
    dispatch(actions.video.updateField({ videoSource: videoUrl }));

    const youTubeId = parseYoutubeId(videoUrl);
    if (youTubeId) {
      dispatch(requests.checkTranscriptsForImport({
        videoId,
        youTubeId,
        onSuccess: (response) => {
          if (response.data.command === 'import') {
            dispatch(actions.video.updateField({
              allowTranscriptImport: true,
            }));
          }
        },
      }));
    }
  },
  updateVideoId: (e) => {
    const updatedVideoId = e.target.value;
    if (previousVideoId !== updatedVideoId && updatedVideoId) {
      setAlert();
    }
    dispatch(actions.video.updateField({ videoId: updatedVideoId }));
  },
});

export const fallbackHooks = ({ fallbackVideos, dispatch }) => ({
  addFallbackVideo: () => dispatch(actions.video.updateField({ fallbackVideos: [...fallbackVideos, ''] })),

  deleteFallbackVideo: (videoIndex) => {
    const updatedFallbackVideos = fallbackVideos.reduce((result, currentItem, index) => {
      if (index === videoIndex) { return result; }
      return [...result, currentItem];
    }, []);

    dispatch(actions.video.updateField({ fallbackVideos: updatedFallbackVideos }));
  },
});

export const videoIdChangeAlert = () => {
  const [showVideoIdChangeAlert, setShowVideoIdChangeAlert] = state.showVideoIdChangeAlert(false);
  return {
    videoIdChangeAlert: {
      show: showVideoIdChangeAlert,
      set: () => setShowVideoIdChangeAlert(true),
      dismiss: () => setShowVideoIdChangeAlert(false),
    },
  };
};

export default {
  videoIdChangeAlert,
  sourceHooks,
  fallbackHooks,
};
