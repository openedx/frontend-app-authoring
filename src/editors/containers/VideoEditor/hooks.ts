import { useState, createContext } from 'react';
import { thunkActions } from '../../data/redux';

import { StrictDict } from '../../utils';

type ErrRecord = Record<string, string>;
type ErrCategory = [errors: ErrRecord, setter: (newErrors: ErrRecord) => void];
interface ErrorContextData {
  duration: ErrCategory;
  handout: ErrCategory;
  license: ErrCategory;
  thumbnail: ErrCategory;
  transcripts: ErrCategory;
  videoSource: ErrCategory;
}

/* istanbul ignore next */
export const ErrorContext = createContext<ErrorContextData>({
  duration: [{}, () => {}],
  handout: [{}, () => {}],
  license: [{}, () => {}],
  thumbnail: [{}, () => {}],
  transcripts: [{}, () => {}],
  videoSource: [{}, () => {}],
});

export const state = StrictDict({
  /* eslint-disable react-hooks/rules-of-hooks */
  durationErrors: (val) => useState(val),
  handoutErrors: (val) => useState(val),
  licenseErrors: (val) => useState(val),
  thumbnailErrors: (val) => useState(val),
  transcriptsErrors: (val) => useState(val),
  videoSourceErrors: (val) => useState(val),
  /* eslint-enable react-hooks/rules-of-hooks */
});

export const errorsHook = (): { error: ErrorContextData, validateEntry: () => boolean } => {
  const [durationErrors, setDurationErrors] = state.durationErrors({});
  const [handoutErrors, setHandoutErrors] = state.handoutErrors({});
  const [licenseErrors, setLicenseErrors] = state.licenseErrors({});
  const [thumbnailErrors, setThumbnailErrors] = state.thumbnailErrors({});
  const [transcriptsErrors, setTranscriptsErrors] = state.transcriptsErrors({});
  const [videoSourceErrors, setVideoSourceErrors] = state.videoSourceErrors({});

  return {
    error: {
      duration: [durationErrors, setDurationErrors],
      handout: [handoutErrors, setHandoutErrors],
      license: [licenseErrors, setLicenseErrors],
      thumbnail: [thumbnailErrors, setThumbnailErrors],
      transcripts: [transcriptsErrors, setTranscriptsErrors],
      videoSource: [videoSourceErrors, setVideoSourceErrors],
    },
    validateEntry: () => {
      if (Object.keys(durationErrors).length > 0) { return false; }
      if (Object.keys(handoutErrors).length > 0) { return false; }
      if (Object.keys(licenseErrors).length > 0) { return false; }
      if (Object.keys(thumbnailErrors).length > 0) { return false; }
      if (Object.keys(transcriptsErrors).length > 0) { return false; }
      if (Object.keys(videoSourceErrors).length > 0) { return false; }
      return true;
    },
  };
};

export const fetchVideoContent = () => ({ dispatch }) => (
  dispatch(thunkActions.video.saveVideoData())
);
