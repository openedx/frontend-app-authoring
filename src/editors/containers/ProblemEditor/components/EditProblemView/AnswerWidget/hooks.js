import { useState, useEffect } from 'react';
import { StrictDict } from '../../../../../utils';
// This 'module' self-import hack enables mocking during tests.
// See src/editors/decisions/0005-internal-editor-testability-decisions.md. The whole approach to how hooks are tested
// should be re-thought and cleaned up to avoid this pattern.
// eslint-disable-next-line import/no-self-import
import * as module from './hooks';
import { actions } from '../../../../../data/redux';
import { ProblemTypeKeys } from '../../../../../data/constants/problem';
import { fetchEditorContent } from '../hooks';

export const state = StrictDict({
  // eslint-disable-next-line react-hooks/rules-of-hooks
  isFeedbackVisible: (val) => useState(val),
});

export const removeAnswer = ({
  answer,
  dispatch,
}) => () => {
  dispatch(actions.problem.deleteAnswer({
    id: answer.id,
    correct: answer.correct,
    editorState: fetchEditorContent({ format: '' }),
  }));
};

export const setAnswer = ({ answer, hasSingleAnswer, dispatch }) => (payload) => {
  dispatch(actions.problem.updateAnswer({ id: answer.id, hasSingleAnswer, ...payload }));
};

export const setAnswerTitle = ({
  answer,
  hasSingleAnswer,
  dispatch,
  problemType,
}) => (updatedTitle) => {
  let title = updatedTitle;
  if ([ProblemTypeKeys.TEXTINPUT, ProblemTypeKeys.NUMERIC, ProblemTypeKeys.DROPDOWN].includes(problemType)) {
    title = updatedTitle.target.value;
  }
  dispatch(actions.problem.updateAnswer({ id: answer.id, hasSingleAnswer, title }));
};

export const setSelectedFeedback = ({ answer, hasSingleAnswer, dispatch }) => (value) => {
  if (value) {
    dispatch(actions.problem.updateAnswer({
      id: answer.id,
      hasSingleAnswer,
      selectedFeedback: value,
    }));
  }
};

export const setUnselectedFeedback = ({ answer, hasSingleAnswer, dispatch }) => (value) => {
  if (value) {
    dispatch(actions.problem.updateAnswer({
      id: answer.id,
      hasSingleAnswer,
      unselectedFeedback: value,
    }));
  }
};

export const useFeedback = (answer) => {
  const [isFeedbackVisible, setIsFeedbackVisible] = module.state.isFeedbackVisible(false);
  useEffect(() => {
    // Show feedback fields if feedback is present
    const isVisible = !!answer.selectedFeedback || !!answer.unselectedFeedback;
    setIsFeedbackVisible(isVisible);
  }, [answer]);

  const toggleFeedback = (open) => {
    // Do not allow to hide if feedback is added
    const { selectedFeedback, unselectedFeedback } = fetchEditorContent({ format: '' });

    if (!!selectedFeedback?.[answer.id] || !!unselectedFeedback?.[answer.id]) {
      setIsFeedbackVisible(true);
      return;
    }
    setIsFeedbackVisible(open);
  };
  return {
    isFeedbackVisible,
    toggleFeedback,
  };
};

export const isSingleAnswerProblem = (problemType) => (
  problemType === ProblemTypeKeys.DROPDOWN
);

export const useAnswerContainer = ({ answers, updateField }) => {
  useEffect(() => {
    let answerCount = 0;
    answers.forEach(answer => {
      if (answer.correct) {
        answerCount += 1;
      }
    });
    updateField({ correctAnswerCount: answerCount });
  }, []);
};

export default {
  state, removeAnswer, setAnswer, setAnswerTitle, useFeedback, isSingleAnswerProblem, useAnswerContainer,
};
