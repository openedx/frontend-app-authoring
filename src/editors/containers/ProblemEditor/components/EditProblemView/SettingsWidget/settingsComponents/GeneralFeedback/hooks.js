import { useState, useEffect } from 'react';
import isEmpty from 'lodash/isEmpty';
import messages from './messages';
// This 'module' self-import hack enables mocking during tests.
// See src/editors/decisions/0005-internal-editor-testability-decisions.md. The whole approach to how hooks are tested
// should be re-thought and cleaned up to avoid this pattern.
// eslint-disable-next-line import/no-self-import
import * as module from './hooks';

export const state = {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  summary: (val) => useState(val),
};

export const generalFeedbackHooks = (generalFeedback, updateSettings) => {
  const [summary, setSummary] = module.state.summary({
    message: messages.noGeneralFeedbackSummary, values: {}, intl: true,
  });

  // eslint-disable-next-line react-hooks/rules-of-hooks
  useEffect(() => {
    if (isEmpty(generalFeedback)) {
      setSummary({ message: messages.noGeneralFeedbackSummary, values: {}, intl: true });
    } else {
      setSummary({
        message: generalFeedback,
        values: {},
        intl: false,
      });
    }
  }, [generalFeedback]);

  const handleChange = (event) => {
    updateSettings({ generalFeedback: event.target.value });
  };

  return {
    summary,
    handleChange,
  };
};
