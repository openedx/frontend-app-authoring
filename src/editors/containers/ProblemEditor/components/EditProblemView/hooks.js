import { useState } from 'react';
import 'tinymce';
import { StrictDict } from '../../../../utils';
import ReactStateSettingsParser from '../../data/ReactStateSettingsParser';
import ReactStateOLXParser from '../../data/ReactStateOLXParser';
import { setAssetToStaticUrl } from '../../../../sharedComponents/TinyMceWidget/hooks';
import { ProblemTypeKeys } from '../../../../data/constants/problem';

export const state = StrictDict({
  // eslint-disable-next-line react-hooks/rules-of-hooks
  isSaveWarningModalOpen: (val) => useState(val),
});

export const saveWarningModalToggle = () => {
  const [isSaveWarningModalOpen, setIsOpen] = state.isSaveWarningModalOpen(false);
  return {
    isSaveWarningModalOpen,
    openSaveWarningModal: () => setIsOpen(true),
    closeSaveWarningModal: () => setIsOpen(false),
  };
};

/** Checks if any tinymce editor in window is dirty */
export const checkIfEditorsDirty = () => {
  const EditorsArray = window.tinymce.editors;
  return Object.entries(EditorsArray).some(([id, editor]) => {
    if (Number.isNaN(parseInt(id, 10))) {
      if (!editor.isNotDirty) {
        return true;
      }
    }
    return false;
  });
};

export const fetchEditorContent = ({ format }) => {
  const editorObject = { hints: [] };
  const EditorsArray = window.tinymce.editors;
  Object.entries(EditorsArray).forEach(([id, editor]) => {
    if (Number.isNaN(parseInt(id, 10))) {
      if (id.startsWith('answer')) {
        const { answers } = editorObject;
        const answerId = id.substring(id.indexOf('-') + 1);
        editorObject.answers = { ...answers, [answerId]: editor.getContent({ format }) };
      } else if (id.includes('Feedback')) {
        const { selectedFeedback, unselectedFeedback, groupFeedback } = editorObject;
        const feedbackId = id.substring(id.indexOf('-') + 1);
        if (id.startsWith('selected')) {
          editorObject.selectedFeedback = { ...selectedFeedback, [feedbackId]: editor.getContent() };
        }
        if (id.startsWith('unselected')) {
          editorObject.unselectedFeedback = { ...unselectedFeedback, [feedbackId]: editor.getContent() };
        }
        if (id.startsWith('group')) {
          editorObject.groupFeedback = { ...groupFeedback, [feedbackId]: editor.getContent() };
        }
      } else if (id.startsWith('hint')) {
        const { hints } = editorObject;
        hints.push(editor.getContent());
      } else {
        editorObject[id] = editor.getContent();
      }
    }
  });
  return editorObject;
};

export const parseState = ({
  problem,
  isAdvanced,
  ref,
  lmsEndpointUrl,
}) => () => {
  const rawOLX = ref?.current?.state.doc.toString();
  const editorObject = fetchEditorContent({ format: '' });
  const reactOLXParser = new ReactStateOLXParser({ problem, editorObject });
  const reactSettingsParser = new ReactStateSettingsParser({ problem, rawOLX });
  const reactBuiltOlx = setAssetToStaticUrl({ editorValue: reactOLXParser.buildOLX(), lmsEndpointUrl });
  return {
    settings: isAdvanced ? reactSettingsParser.parseRawOlxSettings() : reactSettingsParser.getSettings(),
    olx: isAdvanced ? rawOLX : reactBuiltOlx,
  };
};

export const checkForNoAnswers = ({ openSaveWarningModal, problem }) => {
  const simpleTextAreaProblems = [ProblemTypeKeys.DROPDOWN, ProblemTypeKeys.NUMERIC, ProblemTypeKeys.TEXTINPUT];
  const editorObject = fetchEditorContent({ format: '' });
  const { problemType } = problem;
  const { answers } = problem;
  const answerTitles = simpleTextAreaProblems.includes(problemType) ? {} : editorObject.answers;

  const hasTitle = () => {
    const titles = [];
    answers.forEach(answer => {
      const title = simpleTextAreaProblems.includes(problemType) ? answer.title : answerTitles[answer.id];
      if (title.length > 0) {
        titles.push(title);
      }
    });
    if (titles.length > 0) {
      return true;
    }
    return false;
  };

  const hasCorrectAnswer = () => {
    let correctAnswer;
    answers.forEach(answer => {
      if (answer.correct) {
        const title = simpleTextAreaProblems.includes(problemType) ? answer.title : answerTitles[answer.id];
        if (title.length > 0) {
          correctAnswer = true;
        }
      }
    });
    if (correctAnswer) {
      return true;
    }
    return false;
  };

  if (problemType === ProblemTypeKeys.NUMERIC && !hasTitle()) {
    openSaveWarningModal();
    return true;
  }
  if (!hasCorrectAnswer()) {
    openSaveWarningModal();
    return true;
  }
  return false;
};

export const checkForSettingDiscrepancy = ({ problem, ref, openSaveWarningModal }) => {
  const rawOLX = ref?.current?.state.doc.toString();
  const reactSettingsParser = new ReactStateSettingsParser({ problem, rawOLX });
  const problemSettings = reactSettingsParser.getSettings();
  const rawOlxSettings = reactSettingsParser.parseRawOlxSettings();
  let isMismatched = false;

  Object.entries(rawOlxSettings).forEach(([key, value]) => {
    if (value !== problemSettings[key]) {
      isMismatched = true;
    }
  });

  if (isMismatched) {
    openSaveWarningModal();
    return true;
  }
  return false;
};

export const getContent = ({
  problemState,
  openSaveWarningModal,
  isAdvancedProblemType,
  editorRef,
  lmsEndpointUrl,
}) => {
  const problem = problemState;
  const hasNoAnswers = isAdvancedProblemType ? false : checkForNoAnswers({
    problem,
    openSaveWarningModal,
  });
  const hasMismatchedSettings = isAdvancedProblemType ? checkForSettingDiscrepancy({
    ref: editorRef,
    problem,
    openSaveWarningModal,
  }) : false;
  if (!hasNoAnswers && !hasMismatchedSettings) {
    const data = parseState({
      isAdvanced: isAdvancedProblemType,
      ref: editorRef,
      problem,
      lmsEndpointUrl,
    })();
    return data;
  }
  return null;
};
