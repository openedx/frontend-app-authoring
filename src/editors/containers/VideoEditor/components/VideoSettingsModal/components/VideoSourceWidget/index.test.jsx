import 'CourseAuthoring/editors/setupEditorTest';
import React from 'react';
import { dispatch } from 'react-redux';
import { shallow } from '@edx/react-unit-test-utils';

import { formatMessage } from '../../../../../../testUtils';
import { VideoSourceWidgetInternal as VideoSourceWidget } from '.';
import * as hooks from './hooks';

jest.mock('react-redux', () => {
  const dispatchFn = jest.fn();
  return {
    ...jest.requireActual('react-redux'),
    dispatch: dispatchFn,
    useDispatch: jest.fn(() => dispatchFn),
  };
});

jest.mock('../hooks', () => ({
  selectorKeys: ['soMEkEy'],
  widgetValues: jest.fn().mockReturnValue({
    videoId: { onChange: jest.fn(), onBlur: jest.fn(), local: '' },
    videoSource: { onChange: jest.fn(), onBlur: jest.fn(), local: '' },
    fallbackVideos: {
      formValue: ['somEUrL'],
      onChange: jest.fn(),
      onBlur: jest.fn(),
      local: '',
    },
    allowVideoDownloads: { local: false, onCheckedChange: jest.fn() },
    allowVideoSharing: { local: false, onCheckedChange: jest.fn() },
  }),
}));

jest.mock('./hooks', () => ({
  videoIdChangeAlert: jest.fn().mockReturnValue({
    videoIdChangeAlert: {
      set: (args) => ({ set: args }),
      show: false,
      dismiss: (args) => ({ dismiss: args }),
    },
  }),
  sourceHooks: jest.fn().mockReturnValue({
    updateVideoId: (args) => ({ updateVideoId: args }),
    updateVideoURL: jest.fn().mockName('updateVideoURL'),
  }),
  fallbackHooks: jest.fn().mockReturnValue({
    addFallbackVideo: jest.fn().mockName('addFallbackVideo'),
    deleteFallbackVideo: jest.fn().mockName('deleteFallbackVideo'),
  }),
}));

jest.mock('../../../../../../data/redux', () => ({
  selectors: {
    video: {
      allow: jest.fn(state => ({ allowTranscriptImport: state })),
    },
    requests: {
      isFailed: jest.fn(state => ({ isFailed: state })),
    },
  },
}));

describe('VideoSourceWidget', () => {
  const props = {
    // inject
    intl: { formatMessage },
    // redux
    videoSharingEnabledForCourse: false,
  };

  describe('snapshots', () => {
    describe('snapshots: renders as expected with', () => {
      it('default props', () => {
        expect(
          shallow(<VideoSourceWidget {...props} />).snapshot,
        ).toMatchSnapshot();
      });
      it('videoSharingEnabledForCourse=true', () => {
        const newProps = { ...props, videoSharingEnabledForCourse: true };
        expect(
          shallow(<VideoSourceWidget {...newProps} />).snapshot,
        ).toMatchSnapshot();
      });
    });
  });

  describe('behavior inspection', () => {
    let el;
    let hook;
    beforeEach(() => {
      hook = hooks.sourceHooks({ dispatch, previousVideoId: 'someVideoId', setAlert: jest.fn() });
      el = shallow(<VideoSourceWidget {...props} />);
    });
    test('updateVideoId is tied to id field onBlur', () => {
      const expected = hook.updateVideoId;
      expect(el
        // eslint-disable-next-line
        .shallowWrapper.props.children[1].props.children[0].props.children[0]
        .props.onBlur).toEqual(expected);
    });
    test('updateVideoURL is tied to url field onBlur', () => {
      const control = el.shallowWrapper.props.children[1].props.children[1].props.children[0];
      expect(control.props.floatingLabel).toEqual('Video URL');
      control.props.onBlur('onBlur event');
      expect(hook.updateVideoURL).toHaveBeenCalledWith('onBlur event', '');
    });
  });
});
