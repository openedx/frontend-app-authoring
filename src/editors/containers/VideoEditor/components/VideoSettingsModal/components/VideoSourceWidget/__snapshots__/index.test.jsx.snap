// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoSourceWidget snapshots snapshots: renders as expected with default props 1`] = `
<injectIntl(ShimmedIntlComponent)
  fontSize="x-small"
  title="Video source"
>
  <ErrorAlert
    dismissError={[Function]}
    hideHeading={true}
    isError={false}
  >
    <FormattedMessage
      defaultMessage="The Video ID field has changed, please check the Video URL and fallback URL values and update them if necessary."
      description="Body message for the alert that appears when the video id has been changed."
      id="authoring.videoeditor.videoIdChangeAlert.message"
    />
  </ErrorAlert>
  <div
    className="border-primary-100 border-bottom pb-4"
  >
    <Form.Group>
      <Form.Control
        floatingLabel="Video ID"
        onBlur={[Function]}
        onChange={[MockFunction]}
        value=""
      />
      <Form.Control.Feedback
        className="text-primary-300 mb-4"
      >
        <FormattedMessage
          defaultMessage="If you were assigned a video ID by edX, enter the ID here."
          description="Feedback for video ID field"
          id="authoring.videoeditor.videoSource.videoId.feedback"
        />
      </Form.Control.Feedback>
    </Form.Group>
    <Form.Group>
      <Form.Control
        floatingLabel="Video URL"
        onBlur={[Function]}
        onChange={[MockFunction]}
        value=""
      />
      <Form.Control.Feedback
        className="text-primary-300"
      >
        <FormattedMessage
          defaultMessage="The URL for your video. This can be a YouTube URL, or a link
    to an .mp4, .ogg, or .webm video file hosted elsewhere on the internet."
          description="Feedback for video URL field"
          id="authoring.videoeditor.videoSource.videoUrl.feedback"
        />
      </Form.Control.Feedback>
    </Form.Group>
  </div>
  <div
    className="mt-4"
  >
    <FormattedMessage
      defaultMessage="Fallback videos"
      description="Title for the fallback videos section"
      id="authoring.videoeditor.videoSource.fallbackVideo.title"
    />
  </div>
  <div
    className="mt-3"
  >
    <FormattedMessage
      defaultMessage="To be sure all learners can access the video, edX
    recommends providing additional videos in both .mp4 and
    .webm formats.  The first listed video compatible with the
    learner's device will play."
      description="Test explaining reason for fallback videos"
      id="authoring.videoeditor.videoSource.fallbackVideo.message"
    />
  </div>
  <Form.Row
    className="mt-3.5 mx-0 flex-nowrap"
  >
    <Form.Group>
      <Form.Control
        floatingLabel="Video URL"
      />
      <IconButtonWithTooltip
        alt="Delete"
        iconAs="Icon"
        key="top-delete-somEUrL"
        onClick={[Function]}
        tooltipContent="Delete"
        tooltipPlacement="top"
      />
    </Form.Group>
  </Form.Row>
  <ActionRow
    className="mt-4.5"
  >
    <Form.Group>
      <Form.Checkbox
        checked={false}
        className="decorative-control-label"
        onChange={[MockFunction]}
      >
        <div
          className="small text-gray-700"
        >
          <FormattedMessage
            defaultMessage="Allow video downloads"
            description="Label for allow video downloads checkbox"
            id="authoring.videoeditor.videoSource.allowDownloadCheckboxLabel"
          />
        </div>
      </Form.Checkbox>
      <OverlayTrigger
        key="top"
        overlay={
          <Tooltip
            id="tooltip-top"
          >
            <FormattedMessage
              defaultMessage="Allow learners to download versions of this video in
    different formats if they cannot use the edX video player or do not have
    access to YouTube."
              description="Message for allow video downloads checkbox"
              id="authoring.videoeditor.videoSource.allowDownloadTooltipMessage"
            />
          </Tooltip>
        }
        placement="top"
      >
        <Icon
          style={
            {
              "height": "16px",
              "width": "16px",
            }
          }
        />
      </OverlayTrigger>
    </Form.Group>
    <ActionRow.Spacer />
  </ActionRow>
  <div
    className="my-4 border-primary-100 border-bottom"
  />
  <Button
    className="text-primary-500 font-weight-bold pl-0"
    onClick={[Function]}
    size="sm"
    variant="link"
  >
    <FormattedMessage
      defaultMessage="Add a video URL"
      description="Label for add a video URL button"
      id="authoring.videoeditor.videoSource.fallbackVideo.addButtonLabel"
    />
  </Button>
</injectIntl(ShimmedIntlComponent)>
`;

exports[`VideoSourceWidget snapshots snapshots: renders as expected with videoSharingEnabledForCourse=true 1`] = `
<injectIntl(ShimmedIntlComponent)
  fontSize="x-small"
  title="Video source"
>
  <ErrorAlert
    dismissError={[Function]}
    hideHeading={true}
    isError={false}
  >
    <FormattedMessage
      defaultMessage="The Video ID field has changed, please check the Video URL and fallback URL values and update them if necessary."
      description="Body message for the alert that appears when the video id has been changed."
      id="authoring.videoeditor.videoIdChangeAlert.message"
    />
  </ErrorAlert>
  <div
    className="border-primary-100 border-bottom pb-4"
  >
    <Form.Group>
      <Form.Control
        floatingLabel="Video ID"
        onBlur={[Function]}
        onChange={[MockFunction]}
        value=""
      />
      <Form.Control.Feedback
        className="text-primary-300 mb-4"
      >
        <FormattedMessage
          defaultMessage="If you were assigned a video ID by edX, enter the ID here."
          description="Feedback for video ID field"
          id="authoring.videoeditor.videoSource.videoId.feedback"
        />
      </Form.Control.Feedback>
    </Form.Group>
    <Form.Group>
      <Form.Control
        floatingLabel="Video URL"
        onBlur={[Function]}
        onChange={[MockFunction]}
        value=""
      />
      <Form.Control.Feedback
        className="text-primary-300"
      >
        <FormattedMessage
          defaultMessage="The URL for your video. This can be a YouTube URL, or a link
    to an .mp4, .ogg, or .webm video file hosted elsewhere on the internet."
          description="Feedback for video URL field"
          id="authoring.videoeditor.videoSource.videoUrl.feedback"
        />
      </Form.Control.Feedback>
    </Form.Group>
  </div>
  <div
    className="mt-4"
  >
    <FormattedMessage
      defaultMessage="Fallback videos"
      description="Title for the fallback videos section"
      id="authoring.videoeditor.videoSource.fallbackVideo.title"
    />
  </div>
  <div
    className="mt-3"
  >
    <FormattedMessage
      defaultMessage="To be sure all learners can access the video, edX
    recommends providing additional videos in both .mp4 and
    .webm formats.  The first listed video compatible with the
    learner's device will play."
      description="Test explaining reason for fallback videos"
      id="authoring.videoeditor.videoSource.fallbackVideo.message"
    />
  </div>
  <Form.Row
    className="mt-3.5 mx-0 flex-nowrap"
  >
    <Form.Group>
      <Form.Control
        floatingLabel="Video URL"
      />
      <IconButtonWithTooltip
        alt="Delete"
        iconAs="Icon"
        key="top-delete-somEUrL"
        onClick={[Function]}
        tooltipContent="Delete"
        tooltipPlacement="top"
      />
    </Form.Group>
  </Form.Row>
  <ActionRow
    className="mt-4.5"
  >
    <Form.Group>
      <Form.Checkbox
        checked={false}
        className="decorative-control-label"
        onChange={[MockFunction]}
      >
        <div
          className="small text-gray-700"
        >
          <FormattedMessage
            defaultMessage="Allow video downloads"
            description="Label for allow video downloads checkbox"
            id="authoring.videoeditor.videoSource.allowDownloadCheckboxLabel"
          />
        </div>
      </Form.Checkbox>
      <OverlayTrigger
        key="top"
        overlay={
          <Tooltip
            id="tooltip-top"
          >
            <FormattedMessage
              defaultMessage="Allow learners to download versions of this video in
    different formats if they cannot use the edX video player or do not have
    access to YouTube."
              description="Message for allow video downloads checkbox"
              id="authoring.videoeditor.videoSource.allowDownloadTooltipMessage"
            />
          </Tooltip>
        }
        placement="top"
      >
        <Icon
          style={
            {
              "height": "16px",
              "width": "16px",
            }
          }
        />
      </OverlayTrigger>
    </Form.Group>
    <ActionRow.Spacer />
  </ActionRow>
  <div
    className="my-4 border-primary-100 border-bottom"
  />
  <Button
    className="text-primary-500 font-weight-bold pl-0"
    onClick={[Function]}
    size="sm"
    variant="link"
  >
    <FormattedMessage
      defaultMessage="Add a video URL"
      description="Label for add a video URL button"
      id="authoring.videoeditor.videoSource.fallbackVideo.addButtonLabel"
    />
  </Button>
</injectIntl(ShimmedIntlComponent)>
`;
